Allow clients to use a generic random number generator instead of the hard coded OsRng

tonyjzhou opened the pull request:
Try to solve: https://github.com/Trust-Machines/wsts/issues/63

# Tests run locally

## cargo test
```
cargo test
   Compiling wsts v10.0.0 (/Users/tony/Workspace/TrustMachines/wsts)
    Finished `test` profile [unoptimized + debuginfo] target(s) in 1.09s
     Running unittests src/lib.rs (target/debug/deps/wsts-a18cee30445de20f)

running 96 tests
test net::test::signature_share_response_wrapped_verify_msg ... ok
test net::test::dkg_end_verify_msg ... ok
test net::test::signature_share_request_verify_msg ... ok
test net::test::nonce_response_verify_msg ... ok
test net::test::signature_share_response_verify_msg ... ok
test state_machine::coordinator::fire::test::coordinator_state_machine_v1 ... ok
test state_machine::coordinator::fire::test::coordinator_state_machine_v2 ... ok
test net::test::dkg_public_shares_verify_msg ... ok
test net::test::dkg_private_shares_verify_msg ... ok
test net::test::nonce_request_verify_msg ... ok
test state_machine::coordinator::fire::test::equal_after_save_load_v1 ... ok
test net::test::dkg_begin_verify_msg ... ok
test state_machine::coordinator::fire::test::equal_after_save_load_v2 ... ok
test common::test::tuple_proof ... ok
test state_machine::coordinator::fire::test::bad_poly_length_dkg_v2 ... ok
test state_machine::coordinator::fire::test::all_signers_dkg_v2 ... ok
test state_machine::coordinator::fire::test::malicious_signers_dkg_v2 ... ok
test state_machine::coordinator::fire::test::all_signers_sign_v2 ... ok
test state_machine::coordinator::fire::test::bad_poly_length_dkg_v1 ... ok
test state_machine::coordinator::fire::test::malicious_signers_dkg_v1 ... ok
test state_machine::coordinator::fire::test::all_signers_dkg_v1 ... ok
test state_machine::coordinator::fire::test::all_signers_sign_v1 ... ok
2024-11-25T19:12:41.235861Z ERROR wsts::state_machine::coordinator::fire: Timeout gathering nonces for signing round 1 iteration 1, unable to continue
2024-11-25T19:12:41.372979Z ERROR wsts::state_machine::coordinator::fire: Timeout gathering nonces for signing round 1 iteration 1, unable to continue
2024-11-25T19:12:41.807416Z ERROR wsts::state_machine::coordinator::fire: Insufficient non-malicious signers, unable to continue
test state_machine::coordinator::fire::test::insufficient_signers_sign_v2 ... ok
test state_machine::coordinator::fire::test::check_signature_shares_v2 ... ok
2024-11-25T19:12:41.982731Z ERROR wsts::state_machine::coordinator::fire: Insufficient non-malicious signers, unable to continue
test state_machine::coordinator::fire::test::insufficient_signers_sign_v1 ... ok
test state_machine::coordinator::fire::test::new_coordinator_v1 ... ok
test state_machine::coordinator::fire::test::new_coordinator_v2 ... ok
test state_machine::coordinator::fire::test::old_round_ids_are_ignored_v1 ... ok
test state_machine::coordinator::fire::test::old_round_ids_are_ignored_v2 ... ok
2024-11-25T19:12:42.163577Z ERROR wsts::state_machine::coordinator::fire: Timeout gathering DkgPublicShares for dkg round 1 signing round 0 iteration 0, dkg_threshold not met (16/18), unable to continue
2024-11-25T19:12:42.215015Z ERROR wsts::state_machine::coordinator::fire: Timeout gathering DkgPublicShares for dkg round 1 signing round 0 iteration 0, dkg_threshold not met (16/18), unable to continue
2024-11-25T19:12:42.224078Z ERROR wsts::state_machine::coordinator::fire: Timeout gathering DkgPublicShares for dkg round 1 signing round 0 iteration 0, dkg_threshold not met (0/18), unable to continue
2024-11-25T19:12:42.337968Z ERROR wsts::state_machine::coordinator::fire: Timeout gathering DkgPublicShares for dkg round 1 signing round 0 iteration 0, dkg_threshold not met (0/18), unable to continue
2024-11-25T19:12:42.353502Z ERROR wsts::state_machine::coordinator::fire: Timeout gathering DkgPublicShares for dkg round 1 signing round 0 iteration 0, dkg_threshold not met (0/18), unable to continue
test state_machine::coordinator::fire::test::missing_public_keys_dkg_v1 ... ok
2024-11-25T19:12:42.392032Z ERROR wsts::state_machine::coordinator::fire: Timeout gathering DkgPublicShares for dkg round 1 signing round 0 iteration 0, dkg_threshold not met (0/18), unable to continue
test state_machine::coordinator::fire::test::missing_public_keys_dkg_v2 ... ok
test state_machine::coordinator::fire::test::start_dkg_round_v1 ... ok
test state_machine::coordinator::fire::test::start_dkg_round_v2 ... ok
test state_machine::coordinator::fire::test::start_private_shares_v1 ... ok
test state_machine::coordinator::fire::test::start_private_shares_v2 ... ok
test state_machine::coordinator::fire::test::start_public_shares_v1 ... ok
test state_machine::coordinator::fire::test::start_public_shares_v2 ... ok
2024-11-25T19:12:42.446520Z ERROR wsts::state_machine::coordinator::fire: Timeout gathering DkgPublicShares for dkg round 1 signing round 0 iteration 0, dkg_threshold not met (0/18), unable to continue
2024-11-25T19:12:42.508309Z ERROR wsts::state_machine::coordinator::fire: Timeout gathering DkgPublicShares for dkg round 1 signing round 0 iteration 0, dkg_threshold not met (0/18), unable to continue
2024-11-25T19:12:42.562743Z ERROR wsts::state_machine::coordinator::fire: Timeout gathering DkgPublicShares for dkg round 1 signing round 0 iteration 0, dkg_threshold not met (0/18), unable to continue
test state_machine::coordinator::fire::test::multiple_nonce_request_messages_sign_v1 ... ok
test state_machine::coordinator::fire::test::check_signature_shares_v1 ... ok
test state_machine::coordinator::frost::test::coordinator_state_machine_v1 ... ok
test state_machine::coordinator::frost::test::coordinator_state_machine_v2 ... ok
test state_machine::coordinator::fire::test::multiple_nonce_request_messages_sign_v2 ... ok
test state_machine::coordinator::frost::test::equal_after_save_load_v1 ... ok
test state_machine::coordinator::frost::test::new_coordinator_v1 ... ok
test state_machine::coordinator::frost::test::new_coordinator_v2 ... ok
test state_machine::coordinator::frost::test::old_round_ids_are_ignored_v1 ... ok
test state_machine::coordinator::frost::test::old_round_ids_are_ignored_v2 ... ok
test state_machine::coordinator::frost::test::equal_after_save_load_v2 ... ok
2024-11-25T19:12:42.736933Z ERROR wsts::state_machine::coordinator::fire: Timeout gathering DkgPublicShares for dkg round 1 signing round 0 iteration 0, dkg_threshold not met (0/18), unable to continue
test state_machine::coordinator::frost::test::process_inbound_messages_v2 ... ok
test state_machine::coordinator::frost::test::check_signature_shares_v1 ... ok
test state_machine::coordinator::frost::test::start_dkg_round_v1 ... ok
test state_machine::coordinator::frost::test::start_dkg_round_v2 ... ok
test state_machine::coordinator::frost::test::start_private_shares_v1 ... ok
test state_machine::coordinator::frost::test::start_private_shares_v2 ... ok
test state_machine::coordinator::frost::test::start_public_shares_v1 ... ok
test state_machine::coordinator::frost::test::start_public_shares_v2 ... ok
test state_machine::signer::test::can_dkg_end_v1 ... ok
test state_machine::signer::test::can_dkg_end_v2 ... ok
test state_machine::signer::test::dkg_ended_v1 ... ok
test state_machine::signer::test::dkg_ended_v2 ... ok
test state_machine::signer::test::dkg_public_share_v1 ... ok
test state_machine::signer::test::dkg_public_share_v2 ... ok
test state_machine::signer::test::public_shares_done_v1 ... ok
test state_machine::signer::test::public_shares_done_v2 ... ok
test taproot::test::key_tweaks ... ok
test taproot::test::taproot_sign_verify_v1_no_merkle_root ... ok
test state_machine::coordinator::frost::test::run_dkg_sign_v1 ... ok
test taproot::test::taproot_sign_verify_v2_no_merkle_root ... ok
test state_machine::coordinator::frost::test::run_dkg_sign_v2 ... ok
test util::test::test_encrypt_decrypt ... ok
test util::test::test_shared_secret ... ok
2024-11-25T19:12:43.633367Z ERROR wsts::state_machine::coordinator::fire: Timeout gathering DkgPrivateShares for dkg round 1 signing round 0 iteration 0, dkg_threshold not met (16/18), unable to continue
test state_machine::coordinator::fire::test::insufficient_signers_dkg_v2 ... ok
test taproot::test::taproot_sign_verify_v2_with_merkle_root ... ok
test v1::tests::clear_polys ... ok
test taproot::test::taproot_sign_verify_v1_with_merkle_root ... ok
test v1::tests::signer_gen_nonces ... ok
test v1::tests::signer_new ... ok
test v1::tests::signer_save_load ... ok
test v1::tests::run_compute_secrets_missing_shares ... ok
test state_machine::coordinator::frost::test::check_signature_shares_v2 ... ok
test v2::tests::clear_polys ... ok
test v2::tests::party_save_load ... ok
test v2::tests::aggregator_sign ... ok
test v2::tests::run_compute_secrets_missing_shares ... ok
test v1::tests::aggregator_sign ... ok
test v2::tests::bad_polynomial_length ... ok
test v1::tests::bad_polynomial_length ... ok
2024-11-25T19:12:43.840948Z ERROR wsts::state_machine::coordinator::fire: Timeout gathering DkgPrivateShares for dkg round 1 signing round 0 iteration 0, dkg_threshold not met (16/18), unable to continue
test state_machine::coordinator::fire::test::insufficient_signers_dkg_v1 ... ok
test state_machine::coordinator::fire::test::run_dkg_sign_v2 ... ok
test state_machine::coordinator::fire::test::minimum_signers_dkg_v2 ... ok
test state_machine::coordinator::fire::test::minimum_missing_public_keys_sign_v2 ... ok
test state_machine::coordinator::fire::test::minimum_signers_sign_v2 ... ok
test state_machine::coordinator::fire::test::minimum_signers_dkg_v1 ... ok
test state_machine::coordinator::fire::test::minimum_signers_sign_v1 ... ok
test state_machine::coordinator::fire::test::run_dkg_sign_v1 ... ok
test state_machine::coordinator::fire::test::missing_public_keys_sign_v1 ... ok

test result: ok. 96 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 4.43s

     Running unittests src/main.rs (target/debug/deps/wsts-e5dff3b476412678)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

   Doc-tests wsts

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

## cargo clippy
```
cargo clippy -- -D warnings
    Checking wsts v10.0.0 (/Users/tony/Workspace/TrustMachines/wsts)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.71s
```

## cargo bench
```
cargo bench
    Finished `bench` profile [optimized] target(s) in 0.10s
     Running unittests src/lib.rs (target/release/deps/wsts-3fb93162df8059df)

running 96 tests
test common::test::tuple_proof ... ignored
test net::test::dkg_begin_verify_msg ... ignored
test net::test::dkg_end_verify_msg ... ignored
test net::test::dkg_private_shares_verify_msg ... ignored
test net::test::dkg_public_shares_verify_msg ... ignored
test net::test::nonce_request_verify_msg ... ignored
test net::test::nonce_response_verify_msg ... ignored
test net::test::signature_share_request_verify_msg ... ignored
test net::test::signature_share_response_verify_msg ... ignored
test net::test::signature_share_response_wrapped_verify_msg ... ignored
test state_machine::coordinator::fire::test::all_signers_dkg_v1 ... ignored
test state_machine::coordinator::fire::test::all_signers_dkg_v2 ... ignored
test state_machine::coordinator::fire::test::all_signers_sign_v1 ... ignored
test state_machine::coordinator::fire::test::all_signers_sign_v2 ... ignored
test state_machine::coordinator::fire::test::bad_poly_length_dkg_v1 ... ignored
test state_machine::coordinator::fire::test::bad_poly_length_dkg_v2 ... ignored
test state_machine::coordinator::fire::test::check_signature_shares_v1 ... ignored
test state_machine::coordinator::fire::test::check_signature_shares_v2 ... ignored
test state_machine::coordinator::fire::test::coordinator_state_machine_v1 ... ignored
test state_machine::coordinator::fire::test::coordinator_state_machine_v2 ... ignored
test state_machine::coordinator::fire::test::equal_after_save_load_v1 ... ignored
test state_machine::coordinator::fire::test::equal_after_save_load_v2 ... ignored
test state_machine::coordinator::fire::test::insufficient_signers_dkg_v1 ... ignored
test state_machine::coordinator::fire::test::insufficient_signers_dkg_v2 ... ignored
test state_machine::coordinator::fire::test::insufficient_signers_sign_v1 ... ignored
test state_machine::coordinator::fire::test::insufficient_signers_sign_v2 ... ignored
test state_machine::coordinator::fire::test::malicious_signers_dkg_v1 ... ignored
test state_machine::coordinator::fire::test::malicious_signers_dkg_v2 ... ignored
test state_machine::coordinator::fire::test::minimum_missing_public_keys_sign_v2 ... ignored
test state_machine::coordinator::fire::test::minimum_signers_dkg_v1 ... ignored
test state_machine::coordinator::fire::test::minimum_signers_dkg_v2 ... ignored
test state_machine::coordinator::fire::test::minimum_signers_sign_v1 ... ignored
test state_machine::coordinator::fire::test::minimum_signers_sign_v2 ... ignored
test state_machine::coordinator::fire::test::missing_public_keys_dkg_v1 ... ignored
test state_machine::coordinator::fire::test::missing_public_keys_dkg_v2 ... ignored
test state_machine::coordinator::fire::test::missing_public_keys_sign_v1 ... ignored
test state_machine::coordinator::fire::test::multiple_nonce_request_messages_sign_v1 ... ignored
test state_machine::coordinator::fire::test::multiple_nonce_request_messages_sign_v2 ... ignored
test state_machine::coordinator::fire::test::new_coordinator_v1 ... ignored
test state_machine::coordinator::fire::test::new_coordinator_v2 ... ignored
test state_machine::coordinator::fire::test::old_round_ids_are_ignored_v1 ... ignored
test state_machine::coordinator::fire::test::old_round_ids_are_ignored_v2 ... ignored
test state_machine::coordinator::fire::test::run_dkg_sign_v1 ... ignored
test state_machine::coordinator::fire::test::run_dkg_sign_v2 ... ignored
test state_machine::coordinator::fire::test::start_dkg_round_v1 ... ignored
test state_machine::coordinator::fire::test::start_dkg_round_v2 ... ignored
test state_machine::coordinator::fire::test::start_private_shares_v1 ... ignored
test state_machine::coordinator::fire::test::start_private_shares_v2 ... ignored
test state_machine::coordinator::fire::test::start_public_shares_v1 ... ignored
test state_machine::coordinator::fire::test::start_public_shares_v2 ... ignored
test state_machine::coordinator::frost::test::check_signature_shares_v1 ... ignored
test state_machine::coordinator::frost::test::check_signature_shares_v2 ... ignored
test state_machine::coordinator::frost::test::coordinator_state_machine_v1 ... ignored
test state_machine::coordinator::frost::test::coordinator_state_machine_v2 ... ignored
test state_machine::coordinator::frost::test::equal_after_save_load_v1 ... ignored
test state_machine::coordinator::frost::test::equal_after_save_load_v2 ... ignored
test state_machine::coordinator::frost::test::new_coordinator_v1 ... ignored
test state_machine::coordinator::frost::test::new_coordinator_v2 ... ignored
test state_machine::coordinator::frost::test::old_round_ids_are_ignored_v1 ... ignored
test state_machine::coordinator::frost::test::old_round_ids_are_ignored_v2 ... ignored
test state_machine::coordinator::frost::test::process_inbound_messages_v2 ... ignored
test state_machine::coordinator::frost::test::run_dkg_sign_v1 ... ignored
test state_machine::coordinator::frost::test::run_dkg_sign_v2 ... ignored
test state_machine::coordinator::frost::test::start_dkg_round_v1 ... ignored
test state_machine::coordinator::frost::test::start_dkg_round_v2 ... ignored
test state_machine::coordinator::frost::test::start_private_shares_v1 ... ignored
test state_machine::coordinator::frost::test::start_private_shares_v2 ... ignored
test state_machine::coordinator::frost::test::start_public_shares_v1 ... ignored
test state_machine::coordinator::frost::test::start_public_shares_v2 ... ignored
test state_machine::signer::test::can_dkg_end_v1 ... ignored
test state_machine::signer::test::can_dkg_end_v2 ... ignored
test state_machine::signer::test::dkg_ended_v1 ... ignored
test state_machine::signer::test::dkg_ended_v2 ... ignored
test state_machine::signer::test::dkg_public_share_v1 ... ignored
test state_machine::signer::test::dkg_public_share_v2 ... ignored
test state_machine::signer::test::public_shares_done_v1 ... ignored
test state_machine::signer::test::public_shares_done_v2 ... ignored
test taproot::test::key_tweaks ... ignored
test taproot::test::taproot_sign_verify_v1_no_merkle_root ... ignored
test taproot::test::taproot_sign_verify_v1_with_merkle_root ... ignored
test taproot::test::taproot_sign_verify_v2_no_merkle_root ... ignored
test taproot::test::taproot_sign_verify_v2_with_merkle_root ... ignored
test util::test::test_encrypt_decrypt ... ignored
test util::test::test_shared_secret ... ignored
test v1::tests::aggregator_sign ... ignored
test v1::tests::bad_polynomial_length ... ignored
test v1::tests::clear_polys ... ignored
test v1::tests::run_compute_secrets_missing_shares ... ignored
test v1::tests::signer_gen_nonces ... ignored
test v1::tests::signer_new ... ignored
test v1::tests::signer_save_load ... ignored
test v2::tests::aggregator_sign ... ignored
test v2::tests::bad_polynomial_length ... ignored
test v2::tests::clear_polys ... ignored
test v2::tests::party_save_load ... ignored
test v2::tests::run_compute_secrets_missing_shares ... ignored

test result: ok. 0 passed; 0 failed; 96 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running unittests src/main.rs (target/release/deps/wsts-b309afa00fda6bd8)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running benches/v1_bench.rs (target/release/deps/v1_bench-fd9036cdd600a7ab)
Gnuplot not found, using plotters backend
v1 dkg N=20 T=13 K=4    time:   [30.594 ms 30.608 ms 30.623 ms]
                        change: [-1.2317% -1.0782% -0.9280%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 9 outliers among 100 measurements (9.00%)
  6 (6.00%) high mild
  3 (3.00%) high severe

Benchmarking v1 party sign N=20 T=13 K=4: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 9.7s, enable flat sampling, or reduce sample count to 50.
v1 party sign N=20 T=13 K=4
                        time:   [1.9381 ms 1.9414 ms 1.9458 ms]
                        change: [+0.0235% +0.4350% +0.8858%] (p = 0.05 > 0.05)
                        No change in performance detected.
Found 14 outliers among 100 measurements (14.00%)
  4 (4.00%) high mild
  10 (10.00%) high severe

v1 group sign N=20 T=13 K=4
                        time:   [758.55 µs 760.56 µs 762.99 µs]
                        change: [-0.2664% +0.1263% +0.6861%] (p = 0.66 > 0.05)
                        No change in performance detected.
Found 10 outliers among 100 measurements (10.00%)
  5 (5.00%) high mild
  5 (5.00%) high severe

     Running benches/v2_bench.rs (target/release/deps/v2_bench-893ebd1098a1827e)
Gnuplot not found, using plotters backend
v2 dkg N=20 T=13 K=4    time:   [5.9474 ms 5.9571 ms 5.9677 ms]
                        change: [-1.1493% -0.6921% -0.2940%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) high mild
  3 (3.00%) high severe

v2 party sign N=20 T=13 K=4
                        time:   [535.63 µs 537.04 µs 539.16 µs]
                        change: [-0.3954% -0.1413% +0.1162%] (p = 0.32 > 0.05)
                        No change in performance detected.
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) high mild
  3 (3.00%) high severe

v2 group sign N=20 T=13 K=4
                        time:   [91.485 µs 91.599 µs 91.717 µs]
                        change: [-1.1799% -1.0077% -0.8117%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) low mild
  1 (1.00%) high mild
  1 (1.00%) high severe
```

tonyjzhou committed:
Add RNG implementation details and examples to USAGE.md

tonyjzhou committed:
Add dead code allowance for TestConfig struct in net.rs

tonyjzhou committed:
Remove unused imports in benchmark file

tonyjzhou committed:
Remove unnecessary imports from RNG usage examples

tonyjzhou commented:
Hi @xoloki ,

Please kindly review.

xoloki commented on USAGE.md:
```diff
@@ -1,5 +1,44 @@
+### Providing an RNG
+
+You can use the following RNG implementations:
+
```
The text at the bottom referring to the audit needs to come immediately after the discussion of the application’s responsibilities

tonyjzhou replied:
Thanks for catching this. Fixed

xoloki commented:
This is a good start!  Please grep through the code looking for all instances of OsRng both inside and outside of tests. I believe there are some in the state machine message handlers.

xoloki commented on src/v1.rs:
```diff
@@ -860,7 +860,7 @@ mod tests {
         assert_eq!(signer.get_shares().len(), key_ids.len());
         for (_id, shares) in signer.get_shares() {
-            assert_eq!(shares.len(), n.try_into().unwrap());
+            assert_eq!(shares.len(), usize::try_from(n).unwrap());
```
We generally prefer try_into where it is not ambiguous

tonyjzhou replied:
Explicitness and verbosity are the classical trade offs. :)

I will revert it as you have much more context.

tonyjzhou committed:
Fix type conversion for share length validation in tests

tonyjzhou committed:
Clarify usage requirements for packet verification in WSTS

tonyjzhou committed:
Update USAGE.md to add missing newline at end of file

tonyjzhou committed:
Refactor RNG usage to utilize `create_rng` function throughout

tonyjzhou committed:
Replace OsRng with create_rng in benchmark files

tonyjzhou commented:
@xoloki Addressed all review comments and retested:

## cargo clippy

```
cargo clippy -- -D warnings
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.04s
```


## cargo bench

```
cargo bench
   Compiling wsts v10.0.0 (/Users/tony/Workspace/TrustMachines/wsts)
    Finished `bench` profile [optimized] target(s) in 0.79s
     Running unittests src/lib.rs (target/release/deps/wsts-3fb93162df8059df)

running 96 tests
test common::test::tuple_proof ... ignored
test net::test::dkg_begin_verify_msg ... ignored
test net::test::dkg_end_verify_msg ... ignored
test net::test::dkg_private_shares_verify_msg ... ignored
test net::test::dkg_public_shares_verify_msg ... ignored
test net::test::nonce_request_verify_msg ... ignored
test net::test::nonce_response_verify_msg ... ignored
test net::test::signature_share_request_verify_msg ... ignored
test net::test::signature_share_response_verify_msg ... ignored
test net::test::signature_share_response_wrapped_verify_msg ... ignored
test state_machine::coordinator::fire::test::all_signers_dkg_v1 ... ignored
test state_machine::coordinator::fire::test::all_signers_dkg_v2 ... ignored
test state_machine::coordinator::fire::test::all_signers_sign_v1 ... ignored
test state_machine::coordinator::fire::test::all_signers_sign_v2 ... ignored
test state_machine::coordinator::fire::test::bad_poly_length_dkg_v1 ... ignored
test state_machine::coordinator::fire::test::bad_poly_length_dkg_v2 ... ignored
test state_machine::coordinator::fire::test::check_signature_shares_v1 ... ignored
test state_machine::coordinator::fire::test::check_signature_shares_v2 ... ignored
test state_machine::coordinator::fire::test::coordinator_state_machine_v1 ... ignored
test state_machine::coordinator::fire::test::coordinator_state_machine_v2 ... ignored
test state_machine::coordinator::fire::test::equal_after_save_load_v1 ... ignored
test state_machine::coordinator::fire::test::equal_after_save_load_v2 ... ignored
test state_machine::coordinator::fire::test::insufficient_signers_dkg_v1 ... ignored
test state_machine::coordinator::fire::test::insufficient_signers_dkg_v2 ... ignored
test state_machine::coordinator::fire::test::insufficient_signers_sign_v1 ... ignored
test state_machine::coordinator::fire::test::insufficient_signers_sign_v2 ... ignored
test state_machine::coordinator::fire::test::malicious_signers_dkg_v1 ... ignored
test state_machine::coordinator::fire::test::malicious_signers_dkg_v2 ... ignored
test state_machine::coordinator::fire::test::minimum_missing_public_keys_sign_v2 ... ignored
test state_machine::coordinator::fire::test::minimum_signers_dkg_v1 ... ignored
test state_machine::coordinator::fire::test::minimum_signers_dkg_v2 ... ignored
test state_machine::coordinator::fire::test::minimum_signers_sign_v1 ... ignored
test state_machine::coordinator::fire::test::minimum_signers_sign_v2 ... ignored
test state_machine::coordinator::fire::test::missing_public_keys_dkg_v1 ... ignored
test state_machine::coordinator::fire::test::missing_public_keys_dkg_v2 ... ignored
test state_machine::coordinator::fire::test::missing_public_keys_sign_v1 ... ignored
test state_machine::coordinator::fire::test::multiple_nonce_request_messages_sign_v1 ... ignored
test state_machine::coordinator::fire::test::multiple_nonce_request_messages_sign_v2 ... ignored
test state_machine::coordinator::fire::test::new_coordinator_v1 ... ignored
test state_machine::coordinator::fire::test::new_coordinator_v2 ... ignored
test state_machine::coordinator::fire::test::old_round_ids_are_ignored_v1 ... ignored
test state_machine::coordinator::fire::test::old_round_ids_are_ignored_v2 ... ignored
test state_machine::coordinator::fire::test::run_dkg_sign_v1 ... ignored
test state_machine::coordinator::fire::test::run_dkg_sign_v2 ... ignored
test state_machine::coordinator::fire::test::start_dkg_round_v1 ... ignored
test state_machine::coordinator::fire::test::start_dkg_round_v2 ... ignored
test state_machine::coordinator::fire::test::start_private_shares_v1 ... ignored
test state_machine::coordinator::fire::test::start_private_shares_v2 ... ignored
test state_machine::coordinator::fire::test::start_public_shares_v1 ... ignored
test state_machine::coordinator::fire::test::start_public_shares_v2 ... ignored
test state_machine::coordinator::frost::test::check_signature_shares_v1 ... ignored
test state_machine::coordinator::frost::test::check_signature_shares_v2 ... ignored
test state_machine::coordinator::frost::test::coordinator_state_machine_v1 ... ignored
test state_machine::coordinator::frost::test::coordinator_state_machine_v2 ... ignored
test state_machine::coordinator::frost::test::equal_after_save_load_v1 ... ignored
test state_machine::coordinator::frost::test::equal_after_save_load_v2 ... ignored
test state_machine::coordinator::frost::test::new_coordinator_v1 ... ignored
test state_machine::coordinator::frost::test::new_coordinator_v2 ... ignored
test state_machine::coordinator::frost::test::old_round_ids_are_ignored_v1 ... ignored
test state_machine::coordinator::frost::test::old_round_ids_are_ignored_v2 ... ignored
test state_machine::coordinator::frost::test::process_inbound_messages_v2 ... ignored
test state_machine::coordinator::frost::test::run_dkg_sign_v1 ... ignored
test state_machine::coordinator::frost::test::run_dkg_sign_v2 ... ignored
test state_machine::coordinator::frost::test::start_dkg_round_v1 ... ignored
test state_machine::coordinator::frost::test::start_dkg_round_v2 ... ignored
test state_machine::coordinator::frost::test::start_private_shares_v1 ... ignored
test state_machine::coordinator::frost::test::start_private_shares_v2 ... ignored
test state_machine::coordinator::frost::test::start_public_shares_v1 ... ignored
test state_machine::coordinator::frost::test::start_public_shares_v2 ... ignored
test state_machine::signer::test::can_dkg_end_v1 ... ignored
test state_machine::signer::test::can_dkg_end_v2 ... ignored
test state_machine::signer::test::dkg_ended_v1 ... ignored
test state_machine::signer::test::dkg_ended_v2 ... ignored
test state_machine::signer::test::dkg_public_share_v1 ... ignored
test state_machine::signer::test::dkg_public_share_v2 ... ignored
test state_machine::signer::test::public_shares_done_v1 ... ignored
test state_machine::signer::test::public_shares_done_v2 ... ignored
test taproot::test::key_tweaks ... ignored
test taproot::test::taproot_sign_verify_v1_no_merkle_root ... ignored
test taproot::test::taproot_sign_verify_v1_with_merkle_root ... ignored
test taproot::test::taproot_sign_verify_v2_no_merkle_root ... ignored
test taproot::test::taproot_sign_verify_v2_with_merkle_root ... ignored
test util::test::test_encrypt_decrypt ... ignored
test util::test::test_shared_secret ... ignored
test v1::tests::aggregator_sign ... ignored
test v1::tests::bad_polynomial_length ... ignored
test v1::tests::clear_polys ... ignored
test v1::tests::run_compute_secrets_missing_shares ... ignored
test v1::tests::signer_gen_nonces ... ignored
test v1::tests::signer_new ... ignored
test v1::tests::signer_save_load ... ignored
test v2::tests::aggregator_sign ... ignored
test v2::tests::bad_polynomial_length ... ignored
test v2::tests::clear_polys ... ignored
test v2::tests::party_save_load ... ignored
test v2::tests::run_compute_secrets_missing_shares ... ignored

test result: ok. 0 passed; 0 failed; 96 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running unittests src/main.rs (target/release/deps/wsts-b309afa00fda6bd8)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running benches/v1_bench.rs (target/release/deps/v1_bench-fd9036cdd600a7ab)
Gnuplot not found, using plotters backend
v1 dkg N=20 T=13 K=4    time:   [30.471 ms 30.487 ms 30.508 ms]
Found 9 outliers among 100 measurements (9.00%)
  7 (7.00%) high mild
  2 (2.00%) high severe

Benchmarking v1 party sign N=20 T=13 K=4: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 9.7s, enable flat sampling, or reduce sample count to 50.
v1 party sign N=20 T=13 K=4
                        time:   [1.9294 ms 1.9303 ms 1.9317 ms]
Found 12 outliers among 100 measurements (12.00%)
  5 (5.00%) high mild
  7 (7.00%) high severe

v1 group sign N=20 T=13 K=4
                        time:   [752.60 µs 755.03 µs 758.33 µs]
Found 18 outliers among 100 measurements (18.00%)
  4 (4.00%) high mild
  14 (14.00%) high severe

     Running benches/v2_bench.rs (target/release/deps/v2_bench-893ebd1098a1827e)
Gnuplot not found, using plotters backend
v2 dkg N=20 T=13 K=4    time:   [5.8440 ms 5.8492 ms 5.8544 ms]

v2 party sign N=20 T=13 K=4
                        time:   [529.89 µs 530.51 µs 531.37 µs]
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) high mild
  6 (6.00%) high severe

v2 group sign N=20 T=13 K=4
                        time:   [90.323 µs 90.441 µs 90.563 µs]
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high severe
```

tonyjzhou committed:
Refactor RNG implementation to use utility function

xoloki commented on src/state_machine/signer/mod.rs:
```diff
@@ -180,7 +180,7 @@ impl<SignerType: SignerTrait> Signer<SignerType> {
     ) -> Self {
         assert!(threshold <= total_keys);
-        let mut rng = OsRng;
+        let mut rng = create_rng();
```
Instead of calling `create_rng` here, please pass the `rng` in as a parameter to `new`

xoloki commented on src/state_machine/signer/mod.rs:
```diff
@@ -513,7 +513,7 @@ impl<SignerType: SignerTrait> Signer<SignerType> {
 
     fn nonce_request(&mut self, nonce_request: &NonceRequest) -> Result<Vec<Message>, Error> {
-        let mut rng = OsRng;
+        let mut rng = create_rng();
```
Same here, please pass in the `rng`. This will make it necessary to also pass the `rng` in to `process()`, etc.

xoloki commented on src/state_machine/signer/mod.rs:
```diff
@@ -610,7 +610,7 @@ impl<SignerType: SignerTrait> Signer<SignerType> {
 
     fn dkg_begin(&mut self, dkg_begin: &DkgBegin) -> Result<Vec<Message>, Error> {
-        let mut rng = OsRng;
+        let mut rng = create_rng();
```
Same, please pass in `rng`.

xoloki commented on src/state_machine/signer/mod.rs:
```diff
@@ -621,7 +621,7 @@ impl<SignerType: SignerTrait> Signer<SignerType> {
 
     fn dkg_public_begin(&mut self) -> Result<Vec<Message>, Error> {
-        let mut rng = OsRng;
+        let mut rng = create_rng();
```
Here too.

xoloki commented on src/state_machine/signer/mod.rs:
```diff
@@ -654,7 +654,7 @@ impl<SignerType: SignerTrait> Signer<SignerType> {
         dkg_private_begin: &DkgPrivateBegin,
     ) -> Result<Vec<Message>, Error> {
-        let mut rng = OsRng;
+        let mut rng = create_rng();
```
Same.

xoloki commented on src/state_machine/signer/mod.rs:
```diff
@@ -942,8 +941,8 @@ pub mod test {
-        let mut rnd = OsRng;
-        let private_key = Scalar::random(&mut rnd);
+        let mut rng = create_rng();
+        let private_key = Scalar::random(&mut rng);
```
Oh, good catch!

xoloki reviewed (changes requested):
Looks great!  Just need to pass the `rng` into the `Signer` state machine `process` function, and then pass it down to the message handlers.

tonyjzhou committed:
Add RNG parameter to Signer constructor and update tests

tonyjzhou committed:
Refactor signer message processing to use RNG parameter

tonyjzhou commented:
@xoloki Addressed all comments and ready for a revisit 

## cargo clippy
```
cargo clippy -- -D warnings
    Checking wsts v10.0.0 (/Users/tony/Workspace/TrustMachines/wsts)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 1.11s
```

## cargo bench
```
cargo bench
   Compiling proc-macro2 v1.0.92
   Compiling unicode-ident v1.0.14
   Compiling cfg-if v1.0.0
   Compiling libc v0.2.164
   Compiling version_check v0.9.5
   Compiling typenum v1.17.0
   Compiling memchr v2.7.4
   Compiling byteorder v1.5.0
   Compiling serde v1.0.215
   Compiling rustix v0.38.41
   Compiling thiserror v1.0.69
   Compiling regex-syntax v0.8.5
   Compiling glob v0.3.1
   Compiling once_cell v1.20.2
   Compiling bitflags v2.6.0
   Compiling equivalent v1.0.1
   Compiling generic-array v0.14.7
   Compiling autocfg v1.4.0
   Compiling semver v1.0.23
   Compiling aho-corasick v1.1.3
   Compiling clang-sys v1.8.1
   Compiling home v0.5.9
   Compiling hashbrown v0.15.1
   Compiling same-file v1.0.6
   Compiling num-traits v0.2.19
   Compiling winnow v0.6.20
   Compiling indexmap v2.6.0
   Compiling either v1.13.0
   Compiling syn v1.0.109
   Compiling minimal-lexical v0.2.1
   Compiling quote v1.0.37
   Compiling nom v7.1.3
   Compiling syn v2.0.89
   Compiling walkdir v2.5.0
   Compiling regex-automata v0.4.9
   Compiling getrandom v0.2.15
   Compiling rand_core v0.6.4
   Compiling cpufeatures v0.2.16
   Compiling libloading v0.8.5
   Compiling crunchy v0.2.2
   Compiling shlex v1.3.0
   Compiling crossbeam-utils v0.8.20
   Compiling fastrand v2.2.0
   Compiling bindgen v0.64.0
   Compiling errno v0.3.9
   Compiling lazycell v1.3.0
   Compiling radium v0.7.0
   Compiling cexpr v0.6.0
   Compiling rustc-hash v1.1.0
   Compiling log v0.4.22
   Compiling crypto-common v0.1.6
   Compiling inout v0.1.3
   Compiling bitflags v1.3.2
   Compiling lazy_static v1.5.0
   Compiling peeking_take_while v0.1.2
   Compiling which v4.4.2
   Compiling tempfile v3.14.0
   Compiling static_assertions v1.1.0
   Compiling regex v1.11.1
   Compiling subtle v2.6.1
   Compiling cipher v0.4.4
   Compiling universal-hash v0.5.1
   Compiling block-buffer v0.10.4
   Compiling cc v1.2.1
   Compiling toolchain_find v0.4.0
   Compiling itertools v0.10.5
   Compiling ahash v0.8.11
   Compiling tracing-core v0.1.32
   Compiling opaque-debug v0.3.1
   Compiling tap v1.0.1
   Compiling rustc-hex v2.1.0
   Compiling hex v0.4.3
   Compiling regex-syntax v0.6.29
   Compiling wyz v0.5.1
   Compiling crossbeam-epoch v0.9.18
   Compiling uint v0.9.5
   Compiling polyval v0.6.2
   Compiling digest v0.10.7
   Compiling pin-project-lite v0.2.15
   Compiling serde_json v1.0.133
   Compiling overload v0.1.1
   Compiling funty v2.0.0
   Compiling rayon-core v1.12.1
   Compiling tracing-log v0.2.0
   Compiling nu-ansi-term v0.46.0
   Compiling sharded-slab v0.1.7
   Compiling sha2 v0.10.8
   Compiling ghash v0.5.1
   Compiling crossbeam-deque v0.8.5
   Compiling ctr v0.9.2
   Compiling aes v0.8.4
   Compiling bitvec v1.0.1
   Compiling aead v0.5.2
   Compiling thread_local v1.1.8
   Compiling half v2.4.1
   Compiling ciborium-io v0.2.2
   Compiling serde_derive v1.0.215
   Compiling zerocopy-derive v0.7.35
   Compiling thiserror-impl v1.0.69
   Compiling tracing-attributes v0.1.27
   Compiling regex-automata v0.1.10
   Compiling bs58 v0.4.0
   Compiling anstyle v1.0.10
   Compiling ryu v1.0.18
   Compiling itoa v1.0.13
   Compiling allocator-api2 v0.2.20
   Compiling matchers v0.1.0
   Compiling zerocopy v0.7.35
   Compiling smallvec v1.13.2
   Compiling clap_lex v0.7.3
   Compiling tracing v0.1.40
   Compiling plotters-backend v0.3.7
   Compiling clap_builder v4.5.21
   Compiling ciborium-ll v0.2.2
   Compiling ppv-lite86 v0.2.20
   Compiling plotters-svg v0.3.7
   Compiling aes-gcm v0.10.3
   Compiling rand_chacha v0.3.1
   Compiling tracing-subscriber v0.3.18
   Compiling bs58 v0.5.1
   Compiling rand v0.8.5
   Compiling cast v0.3.0
   Compiling plotters v0.3.7
   Compiling fixed-hash v0.8.0
   Compiling primitive-types v0.12.2
   Compiling criterion-plot v0.5.0
   Compiling clap v4.5.21
   Compiling rayon v1.10.0
   Compiling is-terminal v0.4.13
   Compiling oorandom v11.1.4
   Compiling anes v0.1.6
   Compiling serde_spanned v0.6.8
   Compiling toml_datetime v0.6.8
   Compiling polynomial v0.2.6
   Compiling hashbrown v0.14.5
   Compiling ciborium v0.2.2
   Compiling toml_edit v0.22.22
   Compiling tinytemplate v1.2.1
   Compiling criterion v0.5.1
   Compiling toml v0.8.19
   Compiling rustfmt-wrapper v0.2.1
   Compiling p256k1 v7.1.0
   Compiling wsts v10.0.0 (/Users/tony/Workspace/TrustMachines/wsts)
    Finished `bench` profile [optimized] target(s) in 13.79s
     Running unittests src/lib.rs (target/release/deps/wsts-3fb93162df8059df)

running 96 tests
test common::test::tuple_proof ... ignored
test net::test::dkg_begin_verify_msg ... ignored
test net::test::dkg_end_verify_msg ... ignored
test net::test::dkg_private_shares_verify_msg ... ignored
test net::test::dkg_public_shares_verify_msg ... ignored
test net::test::nonce_request_verify_msg ... ignored
test net::test::nonce_response_verify_msg ... ignored
test net::test::signature_share_request_verify_msg ... ignored
test net::test::signature_share_response_verify_msg ... ignored
test net::test::signature_share_response_wrapped_verify_msg ... ignored
test state_machine::coordinator::fire::test::all_signers_dkg_v1 ... ignored
test state_machine::coordinator::fire::test::all_signers_dkg_v2 ... ignored
test state_machine::coordinator::fire::test::all_signers_sign_v1 ... ignored
test state_machine::coordinator::fire::test::all_signers_sign_v2 ... ignored
test state_machine::coordinator::fire::test::bad_poly_length_dkg_v1 ... ignored
test state_machine::coordinator::fire::test::bad_poly_length_dkg_v2 ... ignored
test state_machine::coordinator::fire::test::check_signature_shares_v1 ... ignored
test state_machine::coordinator::fire::test::check_signature_shares_v2 ... ignored
test state_machine::coordinator::fire::test::coordinator_state_machine_v1 ... ignored
test state_machine::coordinator::fire::test::coordinator_state_machine_v2 ... ignored
test state_machine::coordinator::fire::test::equal_after_save_load_v1 ... ignored
test state_machine::coordinator::fire::test::equal_after_save_load_v2 ... ignored
test state_machine::coordinator::fire::test::insufficient_signers_dkg_v1 ... ignored
test state_machine::coordinator::fire::test::insufficient_signers_dkg_v2 ... ignored
test state_machine::coordinator::fire::test::insufficient_signers_sign_v1 ... ignored
test state_machine::coordinator::fire::test::insufficient_signers_sign_v2 ... ignored
test state_machine::coordinator::fire::test::malicious_signers_dkg_v1 ... ignored
test state_machine::coordinator::fire::test::malicious_signers_dkg_v2 ... ignored
test state_machine::coordinator::fire::test::minimum_missing_public_keys_sign_v2 ... ignored
test state_machine::coordinator::fire::test::minimum_signers_dkg_v1 ... ignored
test state_machine::coordinator::fire::test::minimum_signers_dkg_v2 ... ignored
test state_machine::coordinator::fire::test::minimum_signers_sign_v1 ... ignored
test state_machine::coordinator::fire::test::minimum_signers_sign_v2 ... ignored
test state_machine::coordinator::fire::test::missing_public_keys_dkg_v1 ... ignored
test state_machine::coordinator::fire::test::missing_public_keys_dkg_v2 ... ignored
test state_machine::coordinator::fire::test::missing_public_keys_sign_v1 ... ignored
test state_machine::coordinator::fire::test::multiple_nonce_request_messages_sign_v1 ... ignored
test state_machine::coordinator::fire::test::multiple_nonce_request_messages_sign_v2 ... ignored
test state_machine::coordinator::fire::test::new_coordinator_v1 ... ignored
test state_machine::coordinator::fire::test::new_coordinator_v2 ... ignored
test state_machine::coordinator::fire::test::old_round_ids_are_ignored_v1 ... ignored
test state_machine::coordinator::fire::test::old_round_ids_are_ignored_v2 ... ignored
test state_machine::coordinator::fire::test::run_dkg_sign_v1 ... ignored
test state_machine::coordinator::fire::test::run_dkg_sign_v2 ... ignored
test state_machine::coordinator::fire::test::start_dkg_round_v1 ... ignored
test state_machine::coordinator::fire::test::start_dkg_round_v2 ... ignored
test state_machine::coordinator::fire::test::start_private_shares_v1 ... ignored
test state_machine::coordinator::fire::test::start_private_shares_v2 ... ignored
test state_machine::coordinator::fire::test::start_public_shares_v1 ... ignored
test state_machine::coordinator::fire::test::start_public_shares_v2 ... ignored
test state_machine::coordinator::frost::test::check_signature_shares_v1 ... ignored
test state_machine::coordinator::frost::test::check_signature_shares_v2 ... ignored
test state_machine::coordinator::frost::test::coordinator_state_machine_v1 ... ignored
test state_machine::coordinator::frost::test::coordinator_state_machine_v2 ... ignored
test state_machine::coordinator::frost::test::equal_after_save_load_v1 ... ignored
test state_machine::coordinator::frost::test::equal_after_save_load_v2 ... ignored
test state_machine::coordinator::frost::test::new_coordinator_v1 ... ignored
test state_machine::coordinator::frost::test::new_coordinator_v2 ... ignored
test state_machine::coordinator::frost::test::old_round_ids_are_ignored_v1 ... ignored
test state_machine::coordinator::frost::test::old_round_ids_are_ignored_v2 ... ignored
test state_machine::coordinator::frost::test::process_inbound_messages_v2 ... ignored
test state_machine::coordinator::frost::test::run_dkg_sign_v1 ... ignored
test state_machine::coordinator::frost::test::run_dkg_sign_v2 ... ignored
test state_machine::coordinator::frost::test::start_dkg_round_v1 ... ignored
test state_machine::coordinator::frost::test::start_dkg_round_v2 ... ignored
test state_machine::coordinator::frost::test::start_private_shares_v1 ... ignored
test state_machine::coordinator::frost::test::start_private_shares_v2 ... ignored
test state_machine::coordinator::frost::test::start_public_shares_v1 ... ignored
test state_machine::coordinator::frost::test::start_public_shares_v2 ... ignored
test state_machine::signer::test::can_dkg_end_v1 ... ignored
test state_machine::signer::test::can_dkg_end_v2 ... ignored
test state_machine::signer::test::dkg_ended_v1 ... ignored
test state_machine::signer::test::dkg_ended_v2 ... ignored
test state_machine::signer::test::dkg_public_share_v1 ... ignored
test state_machine::signer::test::dkg_public_share_v2 ... ignored
test state_machine::signer::test::public_shares_done_v1 ... ignored
test state_machine::signer::test::public_shares_done_v2 ... ignored
test taproot::test::key_tweaks ... ignored
test taproot::test::taproot_sign_verify_v1_no_merkle_root ... ignored
test taproot::test::taproot_sign_verify_v1_with_merkle_root ... ignored
test taproot::test::taproot_sign_verify_v2_no_merkle_root ... ignored
test taproot::test::taproot_sign_verify_v2_with_merkle_root ... ignored
test util::test::test_encrypt_decrypt ... ignored
test util::test::test_shared_secret ... ignored
test v1::tests::aggregator_sign ... ignored
test v1::tests::bad_polynomial_length ... ignored
test v1::tests::clear_polys ... ignored
test v1::tests::run_compute_secrets_missing_shares ... ignored
test v1::tests::signer_gen_nonces ... ignored
test v1::tests::signer_new ... ignored
test v1::tests::signer_save_load ... ignored
test v2::tests::aggregator_sign ... ignored
test v2::tests::bad_polynomial_length ... ignored
test v2::tests::clear_polys ... ignored
test v2::tests::party_save_load ... ignored
test v2::tests::run_compute_secrets_missing_shares ... ignored

test result: ok. 0 passed; 0 failed; 96 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running unittests src/main.rs (target/release/deps/wsts-b309afa00fda6bd8)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running benches/v1_bench.rs (target/release/deps/v1_bench-fd9036cdd600a7ab)
Gnuplot not found, using plotters backend
v1 dkg N=20 T=13 K=4    time:   [32.015 ms 32.185 ms 32.394 ms]
Found 9 outliers among 100 measurements (9.00%)
  1 (1.00%) high mild
  8 (8.00%) high severe

v1 party sign N=20 T=13 K=4
                        time:   [1.9850 ms 1.9881 ms 1.9912 ms]
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild

v1 group sign N=20 T=13 K=4
                        time:   [771.42 µs 772.68 µs 774.18 µs]
Found 11 outliers among 100 measurements (11.00%)
  2 (2.00%) low mild
  7 (7.00%) high mild
  2 (2.00%) high severe

     Running benches/v2_bench.rs (target/release/deps/v2_bench-893ebd1098a1827e)
Gnuplot not found, using plotters backend
v2 dkg N=20 T=13 K=4    time:   [6.1050 ms 6.1417 ms 6.1861 ms]
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) high mild
  5 (5.00%) high severe

v2 party sign N=20 T=13 K=4
                        time:   [542.18 µs 542.87 µs 543.57 µs]
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) low mild
  2 (2.00%) high mild
  1 (1.00%) high severe

v2 group sign N=20 T=13 K=4
                        time:   [92.728 µs 92.825 µs 92.923 µs]
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) low severe
  3 (3.00%) low mild
  2 (2.00%) high mild
```

xoloki reviewed (approved):
Looks great!  Thanks for taking care of this. 

In the future you don’t need to paste all the passing tests into comments, we can see from CI that they are passing.  But any notable failures, or strangely slow benchmarks would be good to show.
